Add support for exporting and importing provision dialogs

branic opened the pull request:
These rake scripts and classes provide functionality for exporting/importing of the following ManageIQ object types:

- Provision Dialogs

This PR uses the framework that was implemented for PRs #14126, #15256, #16717, and #16983 to export/import other ManageIQ object types.

These scripts are based on the CFME RH Consulting Scripts and are used by Red Hat consultants to enable storing customizations in Git and maintaining customizations between environments (e.g. dev/qa/prod) for an SDLC lifecycle.

Links [Optional]
----------------

* Tracking Issue: [Implement export/import functionality of Miq objects via command-line](#15350)

Steps for Testing/QA [Optional]
-------------------------------
**Exporting**

1. Create a directory for the exports 
```
mkdir /tmp/provision_dialogs
```
2. Export provision dialogs 
```
vmdb
bin/rake evm:export:provision_dialogs -- --directory /tmp/provision_dialogs
```


**Importing**

1. Import all provision dialog yaml files in a directory
```
# vmdb
# bin/rake evm:import:provision_dialogs -- --source /tmp/provision_dialogs
```
2. or Import specific provision_dialog yaml file
```
# vmdb
# bin/rake evm:import:provision_dialogs -- --source /tmp/provision_dialogs/Test_Dialog.yaml
```

branic commented:
/cc @gtanzillo @gmcculloug

gmcculloug commented on lib/task_helpers/exports/provision_dialogs.rb:
```diff
@@ -0,0 +1,27 @@
+        end
+
+        dialogs.each do |dialog|
+          $log.send(:info, "Exporting Provision Dialog: #{dialog[:name]} (#{dialog[:description]})")
```
No need to call `send`.  Use `$log.info`

gmcculloug commented on lib/task_helpers/imports/provision_dialogs.rb:
```diff
@@ -0,0 +1,24 @@
+
+        glob = File.file?(options[:source]) ? options[:source] : "#{options[:source]}/*.yaml"
+        Dir.glob(glob) do |fname|
+          $log.send(:info, "Importing Provision Dialog from: #{fname}")
```
Same here

d-m-u commented on lib/task_helpers/exports/provision_dialogs.rb:
```diff
@@ -0,0 +1,27 @@
+      def export(options = {})
+        export_dir = options[:directory]
+
+        dialogs = if options[:all]
```
This could be a ternary.

d-m-u commented on lib/task_helpers/exports/provision_dialogs.rb:
```diff
@@ -0,0 +1,27 @@
+                    MiqDialog.order(:id).where(:default => [false, nil])
+                  end
+
+        dialogs = dialogs.collect do |dialog|
```
Could use collect! and not need the variable.

branic replied:
`collect!` isn't defined for an `ActiveRecord::Relation []` so the only option is using `collect`

```
     NoMethodError:
       undefined method `collect!' for #<ActiveRecord::Relation []>
       Did you mean?  collect
```

d-m-u commented on lib/task_helpers/imports/provision_dialogs.rb:
```diff
@@ -0,0 +1,24 @@
+
+          miq_dialog = MiqDialog.where(:name => dialog[:name]).first
+
+          if miq_dialog.nil? then
```
This could be a ternary too.

d-m-u replied:
And the ```then``` shouldn't be necessary.

d-m-u commented:
Other than a couple silly nits, looks great! Thanks!

d-m-u commented on lib/task_helpers/imports/provision_dialogs.rb:
```diff
@@ -0,0 +1,24 @@
+
+          dialog = YAML.load_file(fname)
+
+          miq_dialog = MiqDialog.where(:name => dialog[:name]).first
```
Prefer ```find_by(:name => dialog[:name])``` instead of ```where().first```

d-m-u replied:
These are uniq'd by dialog type, does this need to account for that as well?

branic replied:
Both the export and import need to account for them being uniq'd by dialog type. On the export side with the way the filename is generated when there are two dialogs of different types have the same name then one will overwrite the other.

branic commented:
@eclarizio, @d-m-u, @gmcculloug I've addressed the comments with the latest commit. I had to rebase so in the process I squashed the commits into a single commit.

d-m-u commented on lib/task_helpers/exports/provision_dialogs.rb:
```diff
@@ -0,0 +1,22 @@
+
+        dialogs = options[:all] ? MiqDialog.order(:id).all : MiqDialog.order(:id).where(:default => [false, nil])
+
+        dialogs = dialogs.collect do |dialog|
```
You're right about the collect! but a select works here.

gmcculloug replied:
Minor de-duping.  You can move the `order(:all)` down to the caller.

```ruby
dialogs = options[:all] ? MiqDialog.all : MiqDialog.where(:default => [false, nil])

dialogs = dialogs.order(:id).collect do |dialog|
```

gmcculloug commented:
@eclarizio PTAL

branic commented:
@eclarizio With the 2nd `it` (starting on line 85) what I'm trying to test is that two different types of provisioning dialogs with the same name are saved to different files.

I can rework the test to try and be clearer in what is being tested. I'm thinking that I can break the `it` block on line 85 into its own describe block as it really applies regardless of whether the `--all` option was specified.

Thoughts?

eclarizio commented:
@branic Right, I got that after reading it more closely, I was mainly confused at first glance. What you suggested is probably a good way to go about it 👍

branic commented:
@eclarizio does this rework of the spec look ok?

eclarizio commented:
@branic Yeah, that looks fine 👍
